### PR TITLE
Add continuation command implementation

### LINKED ISSUE #133: Continuation command implementation
ℹ️ This issue is part of Epic #131 

## Context

This module implements the `Pact.Command` behaviour. It is in charge of building a command within a continuation payload

## Checklist
- [x] new
- [x] set_network
- [x] set_data
- [x] set_nonce
- [x] set_metadata
- [x] add_keypair
- [x] add_keypairs
- [x] add_signer
- [x] add_signers
- [x] set_pact_tx_hash
- [x] set_step
- [x] set_proof
- [x] set_rollback
- [x] build

### PR DESCRIPTION
**Pull request by EdwinGuayacan**
## Description
Fixes #133 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: lib/pact/command/cont_command.ex [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+  def add_keypair(%__MODULE__{} = cmd_request, keypair) do
+    case KeyPair.new(keypair) do
+      %KeyPair{} = keypair -> add_keypair(cmd_request, keypair)
+      {:error, reason} -> {:error, [keypair: :invalid] ++ reason}
+    end
+  end
+
+  def add_keypair({:error, reason}, _keypair), do: {:error, reason}
+  @impl true
```
Comment (L-Zuluaga) [inline lib/pact/command/cont_command.ex:152]: I think a line break here should be good to separate the implementation from the previous one.

File: lib/pact/command/cont_command.ex [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+    %{cmd_request | pact_tx_hash: pact_tx_hash}
+  end
+
+  def set_pact_tx_hash(%__MODULE__{}, _pact_tx_hash), do: {:error, [pact_tx_hash: :not_a_string]}
+  def set_pact_tx_hash({:error, reason}, _pact_tx_hash), do: {:error, reason}
+
+  @impl true
+  def set_step(%__MODULE__{} = cmd_request, step) when is_integer(step) do
+    %{cmd_request | step: step}
+  end
```
Comment (L-Zuluaga) [inline lib/pact/command/cont_command.ex]: Since the body only has one line, this block could be avoided and instead use the `, do:` notation

File: lib/pact/command/cont_command.ex [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+  def set_step(%__MODULE__{}, _step), do: {:error, [step: :not_an_integer]}
+  def set_step({:error, reason}, _step), do: {:error, reason}
+
+  @impl true
+  def set_proof(%__MODULE__{} = cmd_request, proof) when is_binary(proof) do
+    %{cmd_request | proof: proof}
+  end
```
Comment (L-Zuluaga) [inline lib/pact/command/cont_command.ex]: Same as above

File: lib/pact/command/cont_command.ex [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+  def set_proof(%__MODULE__{}, _proof), do: {:error, [proof: :not_a_string]}
+  def set_proof({:error, reason}, _proof), do: {:error, reason}
+
+  @impl true
+
+  def set_rollback(%__MODULE__{} = cmd_request, rollback) when is_boolean(rollback) do
+    %{cmd_request | rollback: rollback}
+  end
```
Comment (L-Zuluaga) [inline lib/pact/command/cont_command.ex]: Same as above

File: test/pact/command/cont_command_test.exs [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+        sigs: %SignaturesList{
+          signatures: [
+            %Signature{
+              sig:
+                "1c1632e5f98cd7c948c86e90ca7bb984695e70d6dd81fe44d669d71878be3cc11db078cf72ec4cdb67e0fb7ef4730539554c91a559e666c821a6add33c785e01"
+            }
+          ]
+        }
+      } =
+        ContCommand.new(
```
Comment (L-Zuluaga) [inline test/pact/command/cont_command_test.exs]: Maybe we could avoid here the 1-step pipe operator usage

**Review by L-Zuluaga — COMMENTED** [review]
LGTM, just leave a couple of comments about code styles. Excellent work! 🚀

File: lib/pact/command/cont_command.ex [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+    end
+  end
+
+  def set_metadata({:error, reason}, _metadata), do: {:error, reason}
+
+  @impl true
+  def add_keypair(%__MODULE__{keypairs: keypairs} = cmd_request, %KeyPair{} = keypair),
+    do: %{cmd_request | keypairs: keypairs ++ [keypair]}
+
+  def add_keypair(%__MODULE__{} = cmd_request, keypair) do
```
Comment (juanhurtado10) [inline lib/pact/command/cont_command.ex]: @Odraxs @EdwinGuayacan why is this function clause required?

File: lib/pact/command/cont_command.ex [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+          keypairs: keypairs
+        } = cmd_request
+      ) do
+    with {:ok, payload} <- create_payload(cmd_request),
+         {:ok, cmd} <- command_to_json_string(payload, cmd_request),
+         {:ok, sig_commands} <- sign_commands([], cmd, keypairs),
+         {:ok, hash} <- get_unique_hash(sig_commands),
+         {:ok, signatures} <- build_signatures(sig_commands, []),
+         {:ok, command} <- create_command(hash, signatures, cmd) do
+      %Command{} = command
```
Comment (juanhurtado10) [inline lib/pact/command/cont_command.ex]: @EdwinGuayacan @Odraxs IMO, this is not necessary considering that previously `create_command/3` returns a tuple with `{:ok, command}`.

File: lib/pact/command/cont_command.ex [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+  defp get_unique_hash(sign_commands) do
+    unique_hash =
+      Enum.reduce(sign_commands, nil, fn %SignCommand{hash: hash}, acc ->
+        if is_nil(acc) or acc == hash,
+          do: hash,
+          else: {:error, [hash: :hashes_are_not_unique]}
+      end)
+
+    {:ok, unique_hash}
+  end
```
Comment (juanhurtado10) [inline lib/pact/command/cont_command.ex]: @EdwinGuayacan @Odraxs this is duplicate code, which means you have to abstract it from the Command implementations

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: lib/pact/command/cont_command.ex (hunk 6 above)
Comment (EdwinGuayacan) [reply lib/pact/command/cont_command.ex]: So that in case the user wants to set a KeyPair with a raw_data and not with a KeyPair type

Comment (juanhurtado10) [reply lib/pact/command/cont_command.ex]: @EdwinGuayacan Let's remove it, this can lead users to initialization errors, keep it tied to the KeyPair type only.

File: lib/pact/command/cont_command.ex (hunk 8 above)
Comment (Odraxs) [reply lib/pact/command/cont_command.ex]: @juanhurtado10  Does it require a separate utils/helper file, or can it be directly abstracted in the `command behavior`?

Comment (juanhurtado10) [reply lib/pact/command/cont_command.ex]: @Odraxs The Utils/Helper modules have a problem ... all the code that doesn't fit in a module ends up in Utils, so my suggestion is to create a specific module to do this

File: lib/pact/command/cont_command.ex (hunk 6 above)
Comment (EdwinGuayacan) [reply lib/pact/command/cont_command.ex]: the following functions:
set_meta_data,
add_signers,
follow the same logic, do we apply the same modifications?

Comment (juanhurtado10) [reply lib/pact/command/cont_command.ex]: @EdwinGuayacan I'd say yes. The user should initialize them properly before building the Command.

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: lib/pact/command/command_builder.ex [hunk 9]
```diff
@@ -0,0 +1,67 @@
+defmodule Pact.Command.CommandBuilder do
```
Comment (juanhurtado10) [inline lib/pact/command/command_builder.ex]: @EdwinGuayacan @Odraxs Please revert these changes. The word "Builder" is widely used in the OOP paradigm and elixir is a functional technology. And also you are missing the namespace "Kadena" in the module name.

My 5 cents:
1. Create a `Kadena.Pact.Command.Hash` moule.
2. Create a `pull_unique/1` function.
3. Receive a list of signed commands as a parameter.

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: lib/pact/command/command_builder.ex (hunk 9 above)
Comment (Odraxs) [reply lib/pact/command/command_builder.ex]: @juanhurtado10 We leave the other functions in this new module too ? 
Because they are used for both commands ?

Comment (juanhurtado10) [reply lib/pact/command/command_builder.ex]: @Odraxs No, this is a module with a single responsibility. Keep the other functions in their respective command files

File: lib/pact/command/hash.ex [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  @type sign_commands :: list(sign_command())
+  @type valid_hash :: {:ok, hash()}
+
+  @spec pull_unique(sign_commands()) :: valid_hash()
+  def pull_unique(sign_commands) do
+    unique_hash =
+      Enum.reduce(sign_commands, nil, fn %SignCommand{hash: hash}, acc ->
+        if is_nil(acc) or acc == hash,
+          do: hash,
+          else: {:error, [hash: :hashes_are_not_unique]}
```
Comment (juanhurtado10) [inline lib/pact/command/hash.ex]: @Odraxs just a minor tweak and we are ready to merge .. please replace this with `[hash: not_unique]`

**Review by juanhurtado10 — APPROVED** [review]
